### DISCUSSION TITLE
15k TPS with Payload size of 52kb?

### DISCUSSION THREAD

**Discussion by bsandy999** [apache/pulsar#19578] [Q&A]
Hello Folks,
              It's been a month now and I am happy to share my POC experience with the Pulsar Community. Also have some scaling, stability, and configs-related concerns to resolve.

With **Standalone** Setup, I got familiar with the basic concepts/terms and component-wise flow like broker, bookKeeper, and ZooKeeper, etc.

With **Bare Metal** setup, I learned to run Apache Pulsar Cluster on multiple Nodes. Which results in an increase in TPS too.
My Major concern is to achieve maximum TPS with Stable infra, which is missing in my current setup. 

So basically I am doing a load test with a **52kb** payload size with **100 000** records in a single request. To process all **100k** messages it is taking around **46 sec**.

So in short it gives me 4.6k TPS to process 100k records with 52kb payload size each. Where I wanted to achieve min TPS of **_15k with a payload size of 52kb._**

**_Cluster Stability Concern - Benchmark/Load Test Observations (On running load test multiple times)_**

- Brokers get down randomly (sometimes OOM issue, sometimes another random issue like System load, etc)
- Bookie gets down randomly (sometimes OOM issue, sometimes another random issue like System load, etc)
- The data folder does not get a decrease in size even if messages got consumed (persistent storage)
- Broker does not get graceful shutdown on stopping it using CLI (to scale down the number of brokers and to reduce the load on a machine)

All above-mentioned issues are recurring and hence it is impacting POC results.

**_Questions._**

_Kindly suggest the best architecture setup so that setup will be scalable and easy to troubleshoot_
- A basic checklist to maintain while setting up the Pulsar Cluster
- Basic Server Configuration to achieve mentioned TPS
- Cluster tunning & troubleshooting tips to avoid maintenance time

_Sharing the current setup design on which I am running load the test ( All machines are in GCP)_

![image](https://user-images.githubusercontent.com/3489735/220280618-479a5ced-b896-4c4e-9975-1b4b8325866a.png)

Stack Forum -  [Stack Overflow ](https://stackoverflow.com/questions/75518717/apache-pulsar-15k-tps-with-payload-size-of-52kb)

Please go through the above-detailed information and suggest to me the best solution to resolve my requirement  

Thanks in advance!

**Comment by sijie** [comment]
@bsandy999 If I don't mistake the test you did, the TPS is 4.6K per second and the payload is 52KB. The throughput you are getting is 4.6K * 52KB = 239MB / second. That's pretty much what you can get from a 100GB persistent volume on GCP. If you want to support high throughput / tps, you need to look into scaling out the bookkeeper cluster to have more write throughput.

Can you share your configuration for setting up the Pulsar cluster? How many bookies do you have?

**Reply by bsandy999** [reply]
@sijie  Thanks for the feedback. 

So as per the diagram, and as we are running POC only, we have deployed our pulsar cluster on 3 GCP machines, where we are running 1 broker, 1 bookKeeper, and 1 Zk on each machine. So we have scaled our setup horizontally. What if we scale bookies vertically on each machine? So we tried that as well running 3 bookies on a single machine (3 machines * 3 bookies on each machine = in total 9 bookies running ), but with this approach, we faced memory issues and load issues on the system, which causes random bookies shut down an unstable setup. 

**Configuration of Pulsar Instance** 

- Machine 1 - 4 core, 8 GB , 100 GB Disk. **( 1 broker, 1 bookie, 1 zk)**
- Machine 2 - 4 core, 8 GB , 100 GB Disk. **( 1 broker, 1 bookie, 1 zk)**
- Machine 3 - 4 core, 8 GB , 100 GB Disk.  **( 1 broker, 1 bookie, 1 zk)**

This is as per our understanding, if anything seems wrong, then please suggest the best solution where we can utilize each system component in the best manner

**Comment by SushmitaVelan** [comment]
Hi @bsandy999,
I'm really new to the Apache Pulsar and trying to implement a sample POC for Apache Pulsar, a distributed system.
I'm want to establish a communication across 2 machines in which, Machine A has Pulsar Producer C# logic and Machine B has Pulsar Consumer logic.(Note: Both the machines has Pulsar running)
It would be really helpful if you suggest me the steps to implement this above scenario!!
